feat: step2 문자열 계산기

rolroralra opened the pull request:
## 구현
- StringCalculator.kt
- ErrorMessage.kt

## 테스트
- StringCalculatorTests.kt
- StringCalculatorKoTests.kt

rolroralra committed:
test: basic kotest code

rolroralra committed:
test: basic kotest code

rolroralra committed:
feat: step1 문자열 계산기 구현 / 테스트

- StringCalculator.kt, ErrorMessage.kt
- StringCalculatorTests.kt, StringCalculatorKoTests.kt

ddaaac commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,50 @@
+    private val toNumber: (String) -> Int = String::toInt
+) {
+    fun calculate(expression: String?): Int {
+        if (expression.isNullOrBlank()) {
+            throw IllegalArgumentException("${ErrorMessage.NULL_OR_BLANK}")
+        }
```
자바와 코틀린에서 가장 많이 사용하는 두 가지 표준 예외는 `IllegalArgumentException`과 `IllegalStateException`이 있는데요, 코틀린은 두가지 예외에 대한 함수를 제공합니다. [`require(), check()`](https://seosh817.tistory.com/155)에 대해서 알아볼까요?

rolroralra replied:
맞네요. 

`require()`, `check()` 함수 사용하는것이 좋네요.

다른 유용한 함수 있다면 공유 해주시면 감사합니다.

ddaaac commented on src/main/kotlin/step2/ErrorMessage.kt:
```diff
@@ -0,0 +1,10 @@
+package step2
+
+enum class ErrorMessage(val message: String) {
+    NULL_OR_BLANK("Input value is Null or Blank"),
+    INVALID_OPERATOR("Not Supported Operator"),
+    NOT_ENOUGH_OPERAND("Not Enough Operand"),
+    INVALID_NUMBER_FORMAT("Invalid Number Format");
```
어플리케이션이 커져서 에러 메세지가 엄청 많아지면 어떻게 될까요?
각 에러 메세지는 필요한 부분에서 갖고 있으면 좋을 것 같아요.

rolroralra replied:
@ddaaac 
어떻게 하는건지 알려주실수 있을까요?

우선 1차적으로 생각했을 때는,

1. Message Resource (메시지 i18n 까지 고려)
2. Exception 패키지 따로 둬서, 각 Exception Class에 메시지 포함

위 2가지 정도 생각이 납니다 ㅠㅠ

ddaaac commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,50 @@
+import java.lang.NumberFormatException
+
+class StringCalculator(
+    private val plus: String = "+",
+    private val minus: String = "-",
+    private val times: String = "*",
+    private val div: String = "/",
+    private val rem: String = "%",
+    private val spaceCharacter: String = " ",
```
이런 로우값들은 클래스 프로퍼티보다는 [상수](https://www.baeldung.com/kotlin/constants-best-practices)로 정의할 수 있을 것 같습니다 😄

rolroralra replied:
넵.

상수로 변경하겠습니다.

이부분은 사실 연산자 +,-,*,/,%를 클라이언트가 마음대로 변경할수 있다는 구현을 위해서 저렇게 정의하였습니다.

ddaaac commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,50 @@
+
+    private fun calculate(expressionList: List<String>): Int {
+        var accumulator = toValue(expressionList[0])
+        for (i in 1 until expressionList.size step 2) {
+            if (i + 1 >= expressionList.size) {
+                throw IllegalArgumentException("${ErrorMessage.NOT_ENOUGH_OPERAND}")
+            }
```
- depth가 2로 깊어지고 있어요. 이번 교육과정을 진행하시는 동안 각 함수의 depth를 1로 유지해보는 연습을 해보는 건 어떨까요?
- `i + 1 >= size`와 `i > size` 모두 동일한 동작을 할 것 같은데요, "인덱스가 리스트 크기보다 크다"를 표현하는 데는 후자가 조금 더 적절해보여요. 신영님의 생각은 어떠신가요?

rolroralra replied:
넵 depth가 깊어진게 좋지 않은거 같네요.

확실히 SRP를 지키고 역할을 분리하고,

require, check 함수를 이용해서 하는게 좋을거 같네요.

ddaaac commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,50 @@
+    private val div: String = "/",
+    private val rem: String = "%",
+    private val spaceCharacter: String = " ",
+    private val toNumber: (String) -> Int = String::toInt
```
toNumber를 생성자로 받으신 이유가 있을까요?

rolroralra replied:
문자열 연산에서 연산자와 숫자가 구성이 된다고 했을 때,

숫자에 해당하는 String을 Int로 전환하는 걸, 코틀린에서 기본적으로 제공하는걸 쓰되,

만약 이탤리안 숫자나 다른 숫자를 나타내는 문법에 대해서도 구현을 하기 위함이었습니다.

ddaaac commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,50 @@
+            div -> Int::div
+            rem -> Int::rem
+            else -> throw IllegalArgumentException("${ErrorMessage.INVALID_OPERATOR} [$operator]")
+        }
```
when절을 활용해주셨네요 💯 

이 부분을 enum 또는 인터페이스로 분리해보는 건 어떨까요? StringCalculator의 책임을 일부 덜어줄 수 있을 것 같고, 이 부분에 대한 테스트도 쉽게 할 수 있을 것 같습니다 :)

ddaaac commented on src/test/kotlin/step2/StringCalculatorTests.kt:
```diff
@@ -0,0 +1,55 @@
+import org.junit.jupiter.params.provider.ValueSource
+
+class StringCalculatorTests {
+    private var stringCalculator: StringCalculator? = null
+
+    @BeforeEach
+    fun setUp() {
+        stringCalculator = StringCalculator()
+    }
+
+    @ParameterizedTest
+    @CsvSource(delimiterString = "=", value = ["2 + 3 * 4 / 2=10", "2 * 3 / 4 - 2=-1", "2 / 3 + 4 * 2=8"])
+    fun `정상 작동 확인`(expression: String?, expected: Int?) {
+        Assertions.assertThat(stringCalculator!!.calculate(expression)).isEqualTo(expected)
```
[지연-초기화 프로퍼티](https://kotlinlang.org/docs/properties.html#late-initialized-properties-and-variables)에 대해서 알아볼까요?

참고로, non-null assertion 연산자를 `!!`처럼 불편한 모양으로 정한 건, 사용할 때 한 번 더 생각해보고 불편하게 사용하라는 코틀린 개발진들의 의도가 들어가있다고 합니다 😄  최대한 사용을 지양하는 게 좋을 것 같아요.

rolroralra replied:
맞네요. 

코드 리뷰 감사합니다.

ddaaac commented on src/test/kotlin/step2/StringCalculatorKoTests.kt:
```diff
@@ -0,0 +1,91 @@
+            null,
+            " ",
+        ).forEach {
+            "expression: [${it ?: "null"}]" {
```
```suggestion
            "expression: [${it.toString()}]" {
```
같은 결과가 나옵니다. 요 함수는 꼭 내부를 확인해보세요!

rolroralra replied:
감사합니다.

ddaaac commented on src/test/kotlin/step2/StringCalculatorKoTests.kt:
```diff
@@ -0,0 +1,91 @@
+import io.kotest.matchers.string.startWith
+import java.lang.IllegalArgumentException
+
+class StringCalculatorKoTests : FreeSpec({
```
코테스트 와 Junit을 활용해서 테스트를 두 번 작성해주셨군요 💯

rolroralra replied:
코드리뷰 감사합니다

ddaaac reviewed (changes requested):
안녕하세요 신영님 👋 

2단계 구현 잘 해주셨어요 💯 

몇 가지 코멘트를 남겼으니 확인 부탁드려요~

rolroralra committed:
feat: BinaryOperator, ToInt interface segregation

- BinaryOperator(interface), Operator(enum)
- ToInt(interface), StringToInt(implementation)
